### ISSUE TITLE
Connected Components gives wrong results

### ISSUE THREAD

**Issue by wisundstrom** [graphframes/graphframes#453]
TL;DR
I ran across a bug in spark 3.5 and graphframes 0.8.3 where when using connected components, if `spark.sql.adaptive.enabled` is not `false`, the results returned will be incorrect, with many edges being seemingly ignored when finding paths.

  
I've been working on a project that is unfortunately in an air-gapped environment, so I can't share code, but I'll try to provide whatever information I can.

We have been using Graphframes connected components for a few years, and we recently migrated to spark 3.5.0 and graphframes 0.8.3. 
We also migrated from a YARN cluster with hdfs to a k8s cluster with MinIO object storage.

For us, both connected components and BFS are returning results that look like they are not using edges that are present in the edge set.

If we set `spark.sql.adaptive.enabled  =  false`, then the results appear to be calculated as we would have expected.

Mostly wanted to put this here in case anyone else is pulling their hair out over this one, but if anyone has ideas about what could be causing this it would be good to fix, silent errors like this can really be nasty.

**Comment by nchammas** [comment]
Possible duplicate: #428.

**Comment by matt-gorman** [comment]
Thanks for highlighting this @wisundstrom; setting `spark.conf.set('spark.sql.adaptive.enabled', 'false')` alone did not fix this for us using Spark 3.5.0 and Graphframes 0.8.3.

One of the comments on #428 points out that writing and reading edges and vertices from disk worked for them, which this is something that worked for us as well. However, for anyone else running into this where disabling AQE alone doesn't work that it doing to try using written tables, I would point out that initially we were attempting to do this with materialized view output from Delta Live Tables on Databricks, but had the same problems of incorrect output that we had when reading a table and creating edge/vertice DataFrames for Graphframes to consume for connectedComponents, so still had to write those table from the materialized views before reading them into Graphframes.

**Comment by WeichenXu123** [comment]
Hi, could you try GraphFrames 0.8.4 ? it includes a fix https://github.com/graphframes/graphframes/pull/454 , it might fix your issue. would you test it ? :)

**Comment by matt-gorman** [comment]
@WeichenXu123, we tried this with the same result on Databricks using [DBR 15.3 LTS](https://learn.microsoft.com/en-us/azure/databricks/release-notes/runtime/15.4lts-ml), which is Spark 3.5.0 with Graphframes 0.8.4 (0.8.4-db1-spark3.5), with the same incorrect results where edges will end up with different component outputs for their src and dst columns.

Our data comes from a materialized view. If you wanted to try and re-create with test data, our process looks like this.

1. Determine the edges and vertices for each row in a consolidated table (example: `unity.table.consolidated`); these are Arrays of String and a StructType (example below) that holds the src, dst, and relationship of the edges.

```
GraphEdge = T.StructType([
    T.StructField('src',              T.StringType(), True),
    T.StructField('dst',              T.StringType(), True),
    T.StructField('relationship',     T.StringType(), True)
])
```

2. Determine vertices and edges from initial materialized view table:

```
edges = (
    spark.table('unity.table.consolidated')
    .withColumn('EDGES', F.explode(F.col('EDGES')))
    .filter('EDGES IS NOT NULL')
    .select('EDGES.*')
    .distinct()
    .filter('src IS NOT NULL AND dst IS NOT NULL')
)

vertices = (
    spark.table('unity.table.consolidated')
    .select('VERTICES')
    .withColumn('id', F.explode_outer(F.col('VERTICES')))
    .select('id')
    .distinct()
    .filter('id IS NOT NULL')
)
```

4. These are then used in Graphframes to run `connectedComponents` and written to a table:

```
graph = GraphFrame(vertices, edges)

graph_connected = (
    graph.connectedComponents()
)

graph_connected.write.format('delta').mode('overwrite').option('overwriteSchema',True).saveAsTable('unity.table.connected_vertices')
```

5. Our test is joining back the original edges DataFrame by src and dst, grouping by src,dst and checking that number of components per edge is > 1.

This same process works as expected if instead of pushing the edges and vertices DataFrames into a Graphframe, they are first written to a temporary table, then those tables are read into GraphFrame to run connectedComponents.

**Comment by nchammas** [comment]
> This same process works as expected if instead of pushing the edges and vertices DataFrames into a Graphframe, they are first written to a temporary table, then those tables are read into GraphFrame to run connectedComponents.

This was the workaround for #159, before it was fixed by #195.

Perhaps the behavior you are seeing has a similar root cause.

**Comment by WeichenXu123** [comment]
emm, ok. it needs further checking. let me investigate it.

**Comment by olbapjose** [comment]
I am having the same issues with both Databricks graphframes package 0.8.4-db1-spark3.5 and with the open source 0.8.4. 

https://github.com/graphframes/graphframes/issues/159#issuecomment-2386997219
```
spark.conf.set("spark.sql.adaptive.enabled", "false")
g = GraphFrame(v=spark.read.table("catalog.database.nodes"), e=spark.read.table("catalog.database.edges")
```
with no luck (the result contains a massive component with id 0)

**Comment by rjurney** [comment]
See root cause at https://issues.apache.org/jira/browse/SPARK-50992

**Comment by nchammas** [comment]
> See root cause at https://issues.apache.org/jira/browse/SPARK-50992

I don’t follow how the linked Jira issue, which describes an OOM and performance problem, is the root cause of the correctness issue people are discussing here.

**Comment by rjurney** [comment]
> > See root cause at https://issues.apache.org/jira/browse/SPARK-50992
> 
> I don’t follow how the linked Jira issue, which describes an OOM and performance problem, is the root cause of the correctness issue people are discussing here.

Sorry, I think I got my threads crossed. I'll give this a closer look on the morrow...

**Comment by SauronShepherd** [comment]
I've just found out what I think is happening:
- AQE can merge, split, or change the number of partitions at runtime based on data size.
- `monotonically_increasing_id()` assigns IDs based on the original partitions before AQE optimizations. 
- If AQE merges partitions, multiple rows can end up in the same partition with the same ID prefix, causing duplicate IDs.

As suggested in #428, we should replace `monotonically_increasing_id` with `zipWithUniqueId`.
`zipWithUniqueId` is not a DataFrame transformation but an RDD-level transformation, so it is not affected by AQE optimizations.

**NOTE:** Another option could be to temporarily disable the AQE partition coalescing optimization using `spark.config("spark.sql.adaptive.coalescePartitions.enabled", "false").`

**Comment by sonalgoyal** [comment]
I ran into this recently on 3.5.0 at two places, one in the Zingg code before connected components where do some left joins, and another as part of connected components. The left join issue got resolved by moving to 3.5.5, we did this since we found [a known bug with AQE](https://issues.apache.org/jira/browse/SPARK-45282)

I will report back on the connected components part, still runing the tests.

**Comment by sonalgoyal** [comment]
Moving to 3.5.5/3.5.2 started throwing out of memory errors for us. So we are disabling AQE for now. Sorry not a great update :-(

**Comment by SemyonSinchenko** [comment]
> Moving to 3.5.5/3.5.2 started throwing out of memory errors for us. So we are disabling AQE for now. Sorry not a great update :-( 

Does it work for you without AQE?

**Comment by SemyonSinchenko** [comment]
Hi there! As part of #690 we improved CC. From now, passing `broadcastThreshold = -1` will remove the `skewedJoin` (that is the first source of this issue for me) and enable AQE. The second change is that from now CC relies on `DataFrame.checkpoint` (or `DataFrame.localCheckpoint` if checkpoint dir is not set), so no more manual parquet writing and reading (that is the second source of this issue for me). I run some tests and do not see any problems or inconsistency. With `broadcastThreshold = -1` I see also significant performance boost. On the LDBC's wiki-Talks graph, 5M edges, 2M nodes with a gigantic components (the worse case for the algorithm) I see a boost from ~100 seconds to ~20 seconds per iteration.

At the same time, passing positive `broadcastThreshold` (or passing nothing, because the default is 100_000) will work exactly as it was before: without AQE and with manual skewedJoin.

Could be nice if someone can test this changes on big graphs (or provide me some cloud compute resources for such a testing ha-ha). Because of such a performance boost, I'm going to change the default behavior to `broadcastThreshold = -1` and deprecate the old skewedJoin, so it is nice if someone can test it!
